components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,206 @@
+        let mut parent_url = url.clone();
+        if let Ok(mut path_segments) = parent_url.path_segments_mut() {
+            path_segments.pop();
+        }
+        page_html.push_str("<div class=\"parent_link\">");
+        page_html.push_str(&create_link_string("Up to parent directory", &parent_url));
+        page_html.push_str("</div>");
+    }
+
+    page_html.push_str("<div class=\"listing\">");
+    page_html.push_str("<div class=\"header\">");
+    page_html.push_str("<span class=\"name\">Name</span>");
+    page_html.push_str("<span class=\"size\">Size</span>");
+    page_html.push_str("<span class=\"last-modified\">Last Modified</span>");
+    page_html.push_str("</div>");
+
+    for directory_entry in directory_contents {
+        let Ok(directory_entry) = directory_entry else {
+            continue;
+        };
+        let Ok(metadata) = directory_entry.metadata() else {
+            continue;
+        };
+        write_directory_entry(directory_entry, metadata, url, &mut page_html);
+    }
+
+    page_html.push_str("</body>");
+    page_html.push_str("</html>");
+
+    page_html
+}
+
+fn write_directory_entry(entry: DirEntry, metadata: Metadata, url: &Url, output: &mut String) {
+    let Ok(name) = entry.file_name().into_string() else {
+        return;
+    };
+
+    let mut file_url = url.clone();
+    {
+        let Ok(mut path_segments) = file_url.path_segments_mut() else {
+            return;
+        };
+        path_segments.push(&name);
+    }
+
+    let class = if metadata.is_dir() {
+        "directory"
+    } else if metadata.is_symlink() {
+        "symlink"
+    } else {
+        "file"
+    };
+
+    let name = sanitize_html(&name);
+    let link = create_link_string(&name, &file_url);
+    let file_size = metadata_to_file_size_string(&metadata);
+    let last_modified = Utc
+        .timestamp_opt(metadata.mtime(), 0)
+        .single()
+        .map(|time| DateTime::<Local>::from(time))
+        .map(|time| time.format("%F %r").to_string())
+        .unwrap_or_default();
+
+    output.push_str(&format!("<div class=\"entry {class}\">"));
+    output.push_str(&format!("<span class=\"name\">{link}</span>"));
+    output.push_str(&format!("<span class=\"size\">{file_size}</span>"));
+    output.push_str(&format!(
+        "<span class=\"last-modified\">{last_modified}</span>"
+    ));
+    output.push_str("</div>");
+}
+
+fn create_link_string(title: &str, url: &Url) -> String {
+    return format!("<a href=\"{}\">{title}</a>", url.as_str());
+}
+
+pub fn metadata_to_file_size_string(metadata: &Metadata) -> String {
+    if !metadata.is_file() {
+        return String::new();
+    }
+
+    let mut float_size = metadata.size() as f64;
+    let mut prefix_power = 0;
+    while float_size > 1024.0 && prefix_power < 3 {
+        float_size /= 1024.0;
+        prefix_power += 1;
+    }
+
+    let prefix = match prefix_power {
+        0 => "B",
+        1 => "KB",
+        2 => "MB",
+        _ => "GB",
```

Comment:
+1 for [SI binary prefixes](https://physics.nist.gov/cgi-bin/cuu/Info/Units/binary.html) (so KiB, MiB, ...)

Reply:
Good point. I took a look at what different operating system use:

- macOS and Gnome: These use SI units, which means that 1KB is 1000 bytes. This is useful because disk manufacturer *also* use SI units when advertising the size of their disks (to make them seem bigger). Note that for RAM, these operating systems use [https://en.wikipedia.org/wiki/JEDEC_memory_standards#Unit_prefixes_for_semiconductor_storage_capacity](JEDEC) standard units where 1 KB is 1024 bytes.
- Windows uses JEDEC standard units for both RAM and disk so on Windows 1KB is always 1024 bytes.
- Some Linux tools use IEC units (mebibytes, kebibytes, etc), but this is far less common.

I think it's also interesting to point out that both Firefox and Chromium use JEDEC units on my local machine when displaying directory listings. This means the sizes reported are different from those in Nautilus (Gnome Files).

The main issue with SI units is that it's unclear if they are JEDEC units, so on the face of it IEC should be preferred. The big problem with IEC units though is that they are really only known by a small subset of people and the vast majority of computer users do not know what they mean. In addition, it might be confusing if the file sizes differ from what a user sees in their operating system. In addition, it seems like the majority of systems (apart from Windows) simply uses SI units for files and disks.

An ideal situation might be to do what the operating system does, that is to use SI on Linux and macOS or JEDEC on Windows. An alternative would be to match what other browsers do, which is the strategy we take for everything else browser related, focusing on interoperability. In the end, this is pretty minor issue though and I don't like to add platform-specific code (notorious for being tricky to maintain) for something so small. I have chosen to modify this code to use SI units, matching the behavior or macOS, iOS, Gnome, and apparently Android.